relay-server/src/services/spooler/envelope_stack/sqlite.rs

```diff
@@ -0,0 +1,673 @@
+#[derive(Debug, thiserror::Error)]
+pub enum SQLiteEnvelopeStackError {
+    /// The stack is empty.
+    #[error("the stack is empty")]
+    Empty,
+
+    /// The database encountered an unexpected error.
+    #[error("a database error occurred")]
+    DatabaseError(#[from] sqlx::Error),
+}
+
+/// An [`EnvelopeStack`] that is implemented on an SQLite database.
+///
+/// For efficiency reasons, the implementation has an in-memory buffer that is periodically spooled
+/// to disk in a batched way.
+pub struct SQLiteEnvelopeStack {
+    /// Shared SQLite database pool which will be used to read and write from disk.
+    db: Pool<Sqlite>,
+    /// Threshold defining the maximum number of envelopes in the `batches_buffer` before spooling
+    /// to disk will take place.
+    spool_threshold: NonZeroUsize,
+    /// Size of a batch of envelopes that is written to disk.
+    batch_size: NonZeroUsize,
+    /// The project key of the project to which all the envelopes belong.
+    own_key: ProjectKey,
+    /// The project key of the root project of the trace to which all the envelopes belong.
+    sampling_key: ProjectKey,
+    /// In-memory stack containing all the batches of envelopes that either have not been written to disk yet, or have been read from disk recently.
+    #[allow(clippy::vec_box)]
+    batches_buffer: VecDeque<Vec<Box<Envelope>>>,
+    /// The total number of envelopes inside the `batches_buffer`.
+    batches_buffer_size: usize,
+    /// Boolean representing whether calls to `push()` and `peek()` check disk in case not enough
+    /// elements are available in the `batches_buffer`.
+    check_disk: bool,
+}
+
+impl SQLiteEnvelopeStack {
+    /// Creates a new empty [`SQLiteEnvelopeStack`].
+    #[allow(dead_code)]
+    pub fn new(
+        db: Pool<Sqlite>,
+        disk_batch_size: usize,
+        max_batches: usize,
+        own_key: ProjectKey,
+        sampling_key: ProjectKey,
+    ) -> Self {
+        Self {
+            db,
+            spool_threshold: NonZeroUsize::new(disk_batch_size * max_batches)
+                .expect("the spool threshold must be > 0"),
+            batch_size: NonZeroUsize::new(disk_batch_size)
+                .expect("the disk batch size must be > 0"),
+            own_key,
+            sampling_key,
+            batches_buffer: VecDeque::with_capacity(max_batches),
+            batches_buffer_size: 0,
+            check_disk: true,
+        }
+    }
+
+    /// Create a new [`SQLiteEnvelopeStack`] and prefills it with data from disk, up to a fixed
+    /// number of elements.
+    #[allow(dead_code)]
+    pub async fn prepare(
+        db: Pool<Sqlite>,
+        disk_batch_size: usize,
+        max_batches: usize,
+        own_key: ProjectKey,
+        sampling_key: ProjectKey,
+    ) -> Result<(), SQLiteEnvelopeStackError> {
+        let mut stack = Self::new(db, disk_batch_size, max_batches, own_key, sampling_key);
+        stack.unspool_from_disk().await
+    }
+
```

Comment:
nit: Do we call this anywhere? I think we can just lazily load the first batch on pop.

```suggestion
```